mllib/src/main/scala/org/apache/spark/mllib/linalg/Matrices.scala

```diff
@@ -114,21 +106,47 @@ class DenseMatrix(val numRows: Int, val numCols: Int, val values: Array[Double])
   require(values.length == numRows * numCols, "The number of values supplied doesn't match the " +
     s"size of the matrix! values.length: ${values.length}, numRows * numCols: ${numRows * numCols}")
 
-  override def toArray: Array[Double] = values
+  override def toArray: Array[Double] = {
```

Comment:
We promise the users a dense array in column-major ordering. If a user calls `A.transpose.toArray`, that's what we must return. Unfortunately it won't be cheap if it's called on a transposed matrix. Since none of the previous code had `transpose` in it, it won't cause regressions.

Do you think it would make sense to `materialize` the array and reset `isTransposed` if toArray is called, so that if toArray is called again, it won't continuously do the same loop?